File: Serenity.Script.UI/FilterPanel/FilterStore.cs
using Serenity.Data;
using System;
using System.Collections.Generic;
using System.Runtime.CompilerServices;

namespace Serenity
{
    [Imported(ObeysTypeSystem = true)]
    public class FilterStore
    {
        private EventHandler changed;
        private string displayText;

        public FilterStore(List<PropertyItem> fields)
        {
            Items = new List<FilterLine>();

            if (fields == null)
                throw new ArgumentNullException("source");

            this.Fields = fields.Clone();
            this.Fields.Sort((x, y) => Q.Externals.TurkishLocaleCompare(
                Q.TryGetText(x.Title) ?? x.Title ?? x.Name, 
                Q.TryGetText(y.Title) ?? y.Title ?? y.Name));
        
            this.FieldByName = new JsDictionary<string,PropertyItem>();

            foreach (var field in fields)
                FieldByName[field.Name] = field;
        }

        public List<PropertyItem> Fields { get; private set; }
        public JsDictionary<string, PropertyItem> FieldByName { get; private set; }
        public List<FilterLine> Items { get; private set; }

        public void RaiseChanged()
        {
            displayText = null;

            if (changed != null)
                changed(this, EventArgs.Empty);
        }

        public event EventHandler Changed
        {
            add { changed += value; }
            remove { changed -= value; }
        }

        public BaseCriteria ActiveCriteria
        {
            get
            {
                return GetCriteriaFor(Items);
            }
        }

        public static BaseCriteria GetCriteriaFor(List<FilterLine> items)
        {
            if (items == null)
                return Criteria.Empty;

            bool inParens = false;
            BaseCriteria currentBlock = Criteria.Empty;
            bool isBlockOr = false;
            BaseCriteria criteria = Criteria.Empty;

            for (int i = 0; i < items.Count; i++)
            {
                var line = items[i];

                if (line.LeftParen || (inParens && line.RightParen))
                {
                    if (!currentBlock.IsEmpty)
                    {
                        if (inParens)
                            currentBlock = ~(currentBlock);

                        if (isBlockOr)
                            criteria |= currentBlock;
                        else
                            criteria &= currentBlock;

                        currentBlock = Criteria.Empty;
                    }

                    inParens = false;
                }

                if (line.LeftParen)
                {
                    isBlockOr = line.IsOr;
                    inParens = true;
                }

                if (line.IsOr)
                    currentBlock |= line.Criteria;
                else
                    currentBlock &= line.Criteria;
            }

            if (!currentBlock.IsEmpty)
            {
                if (isBlockOr)
                    criteria |= ~(currentBlock);
                else
                    criteria &= ~(currentBlock);
            }

            return criteria;
        }

        public static string GetDisplayTextFor(List<FilterLine> items)
        {
            bool inParens = false;
            var displayText = "";

            if (items == null)
                return displayText;

            for (int i = 0; i < items.Count; i++)
            {
                var line = items[i];

                if (inParens && (line.RightParen || line.LeftParen))
                {
                    displayText += ")";
                    inParens = false;
                }

                if (displayText.Length > 0)
                    displayText += " " + Q.Text("Controls.FilterPanel." + (line.IsOr ? "Or" : "And")) + " ";

                if (line.LeftParen)
                {
                    displayText += "(";
                    inParens = true;
                }

                displayText += line.DisplayText;
            }

            if (inParens)
                displayText += ")";

            return displayText;
        }

        public string DisplayText
        {
            get
            {
                if (displayText == null)
                    displayText = GetDisplayTextFor(Items);

                return displayText;
            }
        }
    }
}

File: Serenity.Scripts/CoreLib/UI/Filtering/FilterStore.ts
namespace Serenity {

    @Decorators.registerClass('FilterStore')
    export class FilterStore {

        constructor(fields: PropertyItem[]) {

            this.items = [];

            if (fields == null) {
                throw new Q.ArgumentNullException('source');
            }

            this.fields = fields.slice();

            this.get_fields().sort(function (x, y) {
                var titleX = Q.tryGetText(x.title);
                if (titleX == null) {
                    titleX = x.title;
                    if (titleX == null)
                        titleX = x.name;
                }

                var titleY = Q.tryGetText(y.title);
                if (titleY == null) {
                    titleY = y.title;
                    if (titleY == null)
                        titleY = y.name;
                }

                return Q.Culture.stringCompare(titleX, titleY);
            });

            this.fieldByName = {};

            for (var field of fields) {
                this.get_fieldByName()[field.name] = field;
            }
        }

        static getCriteriaFor(items: FilterLine[]): any[] {

            if (items == null)
                return [''];

            var inParens = false;
            var currentBlock = [''];
            var isBlockOr = false;
            var criteria = [''];

            for (var i = 0; i < items.length; i++) {
                var line = items[i];

                if (line.leftParen || inParens && line.rightParen) {

                    if (!Serenity.Criteria.isEmpty(currentBlock)) {

                        if (inParens)
                            currentBlock = Serenity.Criteria.paren(currentBlock);

                        if (isBlockOr)
                            criteria = Serenity.Criteria.join(criteria,
                                'or', currentBlock);
                        else
                            criteria = Serenity.Criteria.join(criteria,
                                'and', currentBlock);

                        currentBlock = [''];
                    }

                    inParens = false;
                }

                if (line.leftParen) {
                    isBlockOr = line.isOr;
                    inParens = true;
                }

                if (line.isOr)
                    currentBlock = Serenity.Criteria.join(currentBlock,
                        'or', line.criteria);
                else
                    currentBlock = Serenity.Criteria.join(currentBlock,
                        'and', line.criteria);
            }

            if (!Serenity.Criteria.isEmpty(currentBlock)) {
                if (isBlockOr)
                    criteria = Serenity.Criteria.join(criteria,
                        'or', Serenity.Criteria.paren(currentBlock));
                else
                    criteria = Serenity.Criteria.join(criteria,
                        'and', Serenity.Criteria.paren(currentBlock));
            }

            return criteria;
        }

        static getDisplayTextFor(items: FilterLine[]): string {

            if (items == null)
                return '';

            var inParens = false;
            var displayText = '';

            for (var i = 0; i < items.length; i++) {
                var line = items[i];

                if (inParens && (line.rightParen || line.leftParen)) {
                    displayText += ')';
                    inParens = false;
                }

                if (displayText.length > 0) {
                    displayText += ' ' + Q.text('Controls.FilterPanel.' +
                        (line.isOr ? 'Or' : 'And')) + ' ';
                }

                if (line.leftParen) {
                    displayText += '(';
                    inParens = true;
                }

                displayText += line.displayText;
            }

            if (inParens) {
                displayText += ')';
            }

            return displayText;
        }

        private changed: any;
        private displayText: string;
        private fields: PropertyItem[];
        private fieldByName: Q.Dictionary<PropertyItem>
        private items: FilterLine[];

        get_fields(): PropertyItem[] {
            return this.fields;
        }

        get_fieldByName(): Q.Dictionary<PropertyItem> {
            return this.fieldByName;
        }

        get_items(): FilterLine[] {
            return this.items;
        }

        raiseChanged(): void {
            this.displayText = null;
            this.changed && this.changed(this, {});
        }

        add_changed(value: (e: JQueryEventObject, a: any) => void): void {
            this.changed = Q.delegateCombine(this.changed, value);
        }

        remove_changed(value: (e: JQueryEventObject, a: any) => void): void {
            this.changed = Q.delegateRemove(this.changed, value);
        }

        get_activeCriteria(): any[] {
            return FilterStore.getCriteriaFor(this.items);
        }

        get_displayText(): string {
            if (this.displayText == null)
                this.displayText = FilterStore.getDisplayTextFor(this.items);
            
            return this.displayText;
        }
    }
}